aws/protocol/query/encoder.go

```diff
@@ -1,26 +1,62 @@
 package query
 
-import "net/url"
+import (
+	"io"
+	"net/url"
+	"sort"
+)
 
 // Encoder is a Query encoder that supports construction of Query body
 // values using methods.
 type Encoder struct {
 	// The query values that will be built up to manage encoding.
 	values url.Values
+	// The writer that the encoded body will be written to.
+	writer io.Writer
 	Value
 }
 
 // NewEncoder returns a new Query body encoder
-func NewEncoder() *Encoder {
+func NewEncoder(writer io.Writer) *Encoder {
 	values := url.Values{}
 	return &Encoder{
 		values: values,
+		writer: writer,
 		Value:  newBaseValue(values),
 	}
 }
 
 // Encode returns the []byte slice representing the current
 // state of the Query encoder.
-func (e Encoder) Encode() ([]byte, error) {
-	return []byte(e.values.Encode()), nil
+func (e Encoder) Encode() error {
+	// Get the keys and sort them to have a stable output
+	keys := make([]string, 0, len(e.values))
+	for k := range e.values {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	isFirstEntry := true
+	for _, key := range keys {
+		queryValues := e.values[key]
+		escapedKey := url.QueryEscape(key)
+		for _, value := range queryValues {
+			if !isFirstEntry {
+				if _, err := e.writer.Write([]byte(`&`)); err != nil {
+					return err
+				}
+			} else {
+				isFirstEntry = false
+			}
+			if _, err := e.writer.Write([]byte(escapedKey)); err != nil {
+				return err
+			}
+			if _, err := e.writer.Write([]byte(`=`)); err != nil {
+				return err
+			}
+			if _, err := e.writer.Write([]byte(url.QueryEscape(value))); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
```

Comment:
To simplify this you could continue to use the `url.Value.Encode` method, and write the output of that to the `writer` the Query `Encoder` was created with.  Can also optimize for the `WriteString` method that many writer implementations implement, e.g. [byte.Buffer.WriteString](https://pkg.go.dev/bytes?tab=doc#Buffer.WriteString).

```go
func (e *Encoder) Encode() error {
     encoded := e.values.Encode()
     if ws, ok := e.writer.(interface{ WriteString(string) (int, error) }); ok {
         // Optimize for WriteString to prevent additional byte array being allocated.
         return ws.WriteString(encoded)
     }

     return e.writer.Write([]byte(encoded))
}
```

The idea behind taking an `io.Writer` as input is that later, if we want to improve performance of the query `Encoder` the utility would be writing to a stream of bytes instead of using the `url.Values` utility. `url.Values` as an intermediate encoding datatype will cause unneeded extra allocations per member written.

Reply:
Not deferring to `url.Values.Encode` was a suggestion from @skotambkar to eliminate the extra allocation from the buffer it creates. Is that just not much of a concern here, or is there something else I'm not seeing?

Reply:
An i see. In that case it would be good to remove the need for the extra `[]byte` casts ,because there is a good chance each one of those will do an allocation, because a byte slice is mutable.

One of way to address this is to test if the `io.Writer` passed into `Encoder`'s constructor has the `WriteString` method up front, and if not wrap it with a type that does for convenience.

```go
type wrapWriterString struct {
      w io.Writer
}
func (w wrapWriterString) WriteString(v string) (int, error) {
      return w.Write([]byte(v))
}
```

Then can update the `Encode` method to check for a string writer, and if not wrap the writer.  The SDK should always be using an io.Writer implementation that has the `WriteString` method, e.g. `bytes.Buffer`

```go
func (e Encoder) Encode() error {
     ws, ok := e.writer.(interface{ WriteString(string) (int, error) })
     if !ok {
          // Falls back to less optimal byte slice casting if not available.
          ws = &wrapWriteString{w: e.writer}
     }

     // ...
     //  use as string writer, ws.WriteString(escapedKey)
     // ...
}

Reply:
👍